include/psen_scan/async_udp_reader.h

```diff
@@ -0,0 +1,257 @@
+// Copyright (c) 2020 Pilz GmbH & Co. KG
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+#ifndef PSEN_SCAN_ASYNC_UDP_READER_H
+#define PSEN_SCAN_ASYNC_UDP_READER_H
+
+#include <stdexcept>
+#include <string>
+#include <thread>
+#include <atomic>
+#include <functional>
+#include <iostream>
+
+#include <boost/asio.hpp>
+#include <boost/bind.hpp>
+
+#include "psen_scan/open_connection_failure.h"
+#include "psen_scan/close_connection_failure.h"
+
+namespace psen_scan
+{
+class InvalidDataHandler : public std::invalid_argument
+{
+public:
+  InvalidDataHandler(const std::string& msg = "New data handler is invalid");
+};
+
+class InvalidErrorHandler : public std::invalid_argument
+{
+public:
+  InvalidErrorHandler(const std::string& msg = "Error handler is invalid");
+};
+
+template <std::size_t NumberOfBytes>
+using NewDataHandler = std::function<void(const std::array<char, NumberOfBytes>&, const std::size_t&)>;
+
+using ErrorHandler = std::function<void(const std::string&)>;
+
+static const boost::posix_time::seconds DEFAULT_TIMEOUT{ 1 };
+
+template <std::size_t NumberOfBytes>
```

Comment:
Please reconsider this template. I don't think we will have multiple template instantiations. Especially since this value represents that max-size of a UDP message.

Reply:
I postpone this decision until everything is merged on the feature/new_firmeware branch, in the hope of a more information when everything is integrated.